### PR TITLE
typerel: rework `tyGenericInvocation` handling

### LINKED ISSUE #846: simplify instantiation of generic types (`semtypinst`)
## Goal

Have `semtypinst` only produce fully resolved non-meta types (i.e., types that don't contain any unresolved generic parameters, type-classes, `any`, etc.).

## Why?

At present, the type parameter application implemented in `semtypinst` supports two modes: with meta-types and without (signaled by the `allowMetaTypes` option). The largest differences between both modes is that in "meta-types allowed" mode:
- simple copies are created (via `exactReplica`) instead of proper ones
- expressions in types (ranges, dependent types, etc.) are skipped
- no types are committed to the instantiation cache
- interaction with `ModuleGraph` state (e.g., hooks) is disabled

A different way to look at the "allow meta-type" mode is that it allows for replacing type variables while some of them may be still unknown.

Two problems that arise from this:
1. there a two paths through `semtypinst`, which makes the code very non-linear, and thus hard to follow
2. the types coming out of `semtypinst` in "meta-types allowed" are *improper* ones, whose presence complicates other things (for example, the `sameObjectType` logic, which now can't rely on ID comparisons)

The reason for meta-`tyGenericInst` types existing seems to mainly be for the convenience of `typeRel`, which lacks proper support for unapplied `tyGenericInvocation`s (a `tyGenericInst` is effectively an evaluated `tyGenericInvocation`).

## Concrete benefits

Not having the "allow meta-type" mode would significantly simplify the logic in `semtypinst` and make the code easier to reason about. In addition:
- `tyGenericInst` would change to always represent fully resolved non-meta types
- while some `exactReplica` usages are likely going to be still required, a lot of them can removed

Both, in turn, would either help with or make possible further improvements to `semtypinst` and other type-related fixes.

## How?

Removing the "meta-types allowed mode" will first require moving all of its users. These are (in no specific order):
- [x] `instGenericContainer` usage when turning a `tyGenericBody` into a `tyCompositeTypeClass`. Done: #864 
- [x] `instGenericContainer` usage for instantiating generic concepts. Done: #861
- [x] `prepareMetatypeForSigmatch` in `tryGenerateInstance`. Done: #858
- [x] `prepareMetatypeForSigmatch` in `typeRel`. Done: #854
- [x] `replaceTypesInBody` in `tryResolvingStaticExpr`. Done: #850

Many of the aforementioned boil down to duplicating and adjusting the relevant `replaceTypeVarsT` and `handleGenericInvocation` parts into standalone procedures.

### PR DESCRIPTION
**Pull request by zerbina**
## Summary

Make the type relationship computation (`typeRel`) for instantiated
generic types (`tyGenericInst`) and unresolved type applications
(`tyGenericInvocation`) more consistent with each other, which, apart
from removing another dependency on `semtypinst`'s "meta-types allowed"
mode, is also a preparation for fixing the handling of phantom types.

**Fixes:**
- invocations of types like `Body[T] = T.param` and
  `Body[N] = range[0..(N+1)]` with generic parameters in routine
  signatures now work properly during parameter matching
- for the procedure with the signature `proc f[T](x: Type[Type[T]])`
  and `Type[A] = (A,)`, inferring `T` as `X` now works when passing
  a `(Type[X],)`

A temporary regression is that constraints of generic parameters are now
ignored when a non-generic-instance is matched against the invocation of
a generic type.

## Details

**Key changes:**
- `typeRel` doesn't directly use `prepareMetatypeForSigmatch` anymore,
  which is another step towards removing `semtypinst`'s "meta-types
  allowed" mode
- the `replaceTypeVarsInType` procedure is added, which replaces type
  variables without also resolving `tyGenericInvocation`s 
- `tyGenericInvocation` handling in `typeRel` is now much more similar
  to that of `tyGenericInst`

When computing the relationship with a `tyGenericInvocation` (e.g., a
`Type[T]` in a routine signature where `T` is a generic parameter) and:
- the argument type is an array type or an instantiated generic type
  (i.e., `tyGenericInst`)
- `typeRel` happens in the context of routine signature parameter type
  matching

the invocation was first resolved via `prepareMetatypeForSigmatch` and
the argument type then matched against the instantiated type. The reason
for this seems to have been being able to consistently handle generic
aliases (by skipping them) and, in the case of `array`s, that static
parameter inference works.

What `prepareMetatypeForSigmatch` does is create a fully instantiated
`tyGenericInst` for a `tyGenericInvocation`, using the "meta-types
allowed" mode of `semtypinst`. For example, the instance produced for
`Body[A] = (A, A)` with an invocation like `Body[T]` would be a `(T, T)`
type.

### Reworked handling

In order to not having to rely on the "meta-types allowed" mode, the
handling of `tyGenericInvocation` is reworked to be the following (`f`
is the formal type and `a` the argument type after generic alias are
skipped):
1. if `a` is a `tyGenericInvocation`, the behaviour is the same as
   previously
2. if `a` is a `tyGenericInst` and it's an instance of the generic type
   that the invocation invokes, their relation is that of their
   arguments (the comparison works the same as for two `tyGenericInst`)
3. if the formal invocation is of an `object` type (including `ref` and
   `ptr` versions), `f` and `a` have a sub-type relationship if `a` is a
   sub-type of `f`, otherwise there's no relation
4. if `f` is an invocation of a `and|or` type or of another
   invocation and `a` is a `tyGenericInst`, there is no relationship
   (this mirrors the `tyGenericInst`-against-`tyGenericInst` handling)
5. otherwise (i.e., none of the above were true), the type parameters in
   `f`'s body are replaced with the types passed to the invocation, and
   `a` is matched against the resulting type

The replacing in step 5. is implemented by the new
`replaceTypeParamsInType` procedure, which is similar to
`replaceTypeVarsT`, but with the important difference that:
- it doesn't instantiate `tyGenericInvocation`s
- it properly handles `tyFromExpr`, by also replacing the type
  parameters in the attached AST
- the attached AST for object, tuple, and proc types is not modified, as
  this is not needed for `typeRel` to work with the types

For the case where `a` is not a `tyGenericInst` or `tyArray`, the
important difference is that `a` is now not matched against the body of
`f` first, with the inferred parameters of `f` then being matched
against its arguments. This leads to inference working in more complex
cases, but it also means that the parameter constraints of the invoked
types are ignored. For example:

```nim
type Type[A: object] = (A,)
proc f[T](x: Type[T]) = discard

f((1,))
```
doesn't fail now, while it previously (correctly) did.

**Related changes:**
- move the logic for comparing the arguments of two `tyGenericInst`
  types into a standalone procedure and adjust it to also work with
  `tyGenericInvocation`s
- change `skipToObject` to not skip over `tyGenericBody`, as doing
  so is almost never what is wanted
- adjust `isSubtypeOfGenericInstance` to support `f` being a
  `tyGenericInvocation`
- remove the now-obsolete `isGenericSubtype`
- add a known-issue test for a pre-existing issue with generic range
  types

### CODE REVIEW

File: compiler/sem/sigmatch.nim [hunk 1]
```diff
@@ -1597,88 +1593,70 @@ typeRel can be used to establish various relationships between types:
...
+      # a resolved type application is matched against an unresolved one where
+      # both are for the same generic type -> the arguments have to match
+      result = compareInvocationArguments(c, f, x, flags)
+    elif x.kind == tyGenericBody and f.base == x:
+      # this is a special case to support:
+      #   proc f[T](a: typedesc[Generic[T]])
+      #   f[int](Generic)
+      # XXX: this seems like a really bad idea, why should that work? The
+      #      procedure wants a typedesc for an instantiated type, but it gets
+      #      an (unlifted) composite-type-class
```
Comment (saem) [inline compiler/sem/sigmatch.nim:1622]: It shouldn't, pretty sure it should be an error where `Generic[int]` is expected, but received a `Generic`.

Specifying the type parameter on the procedure should affect the formal parameter's type definition and interpretation of any argument (the instantiated procedure's declaration), but not the actual argument.

File: compiler/sem/sigmatch.nim [hunk 2]
```diff
@@ -1597,88 +1593,70 @@ typeRel can be used to establish various relationships between types:
...
+      # ``Body[A]``, the resulting type would be ``(A, A)``.
+      var bindings: TIdTable
+      initIdTable(bindings)
+      # prepare the parameter-to-argument bindings
+      for i in 1..<f.len:
+        idTablePut(bindings, f.base[i-1], f[i])
+
+      # XXX: since we're forwarding the arguments here, the constraints on the
+      #      body's parameters are ignored...
+      let other = replaceTypeParamsInType(c.c, bindings, body)
```
Comment (saem) [inline compiler/sem/sigmatch.nim:1658]: It's a temporary regression because, it's not unreasonable to do the check as part of the replace in the future?

File: tests/lang_callable/generics/tgeneric_range_invocation_issue.nim [hunk 3]
```diff
@@ -0,0 +1,24 @@
+discard """
+  description: '''
+    Invoking a generic range type with a ``static`` normal parameter instead
+    of a generic parameter must work
+  '''
+  knownIssue: '''
+    The type variable in the range expression reaches the final
+    ``semtypinst.prepareNode`` call as something that is not detected as a
+    type variables
```
Comment (saem) [inline tests/lang_callable/generics/tgeneric_range_invocation_issue.nim:9]:
```suggestion
    type variable
```

File: tests/lang_callable/generics/tgeneric_range_invocation_issue.nim [hunk 4]
```diff
@@ -0,0 +1,24 @@
...
+
+proc f(A: static int, rng: Range[A]): int =
+  result = typeof(rng).high.int
+
+# test with non-range:
+doAssert f(3, 0) == 5
+# test with range:
+var a: range[7..10]
+doAssert f(3, a) == 8
+# negative test: make sure that range types not overlapping are rejected
```
Comment (saem) [inline tests/lang_callable/generics/tgeneric_range_invocation_issue.nim:23]: these are super valuable, nice!

**Review by saem — APPROVED** [review]
I agree the temporary regression is acceptable.

Found a non-blocking very minor typo, I wouldn't bother fixing it -- could easily be done in a future PR, and since it's in 'known issue' text might be removed entirely.

File: compiler/sem/sigmatch.nim (hunk 2 above)
Comment (zerbina) [reply compiler/sem/sigmatch.nim:1658]: Yep, my current plan goes into that direction. In short, the idea is to, should the `typeRel` against the patched body return a match, perform a `typeRel` for all type arguments against their passed-to parameter's constraint. For that to work, `typeRel` is going to need some more improvements, which is why I haven't implemented it as part of this PR.

---

A complementary improvement could be leveraging the `tyInferred` mechanism (or an improved/extended version thereof) to infer the constraints for a generic parameter on definition:
```nim
type Generic[A: object] = (A, A)

proc f[T](x: Generic[T]) = ...
```
here `T` would be inferred to be constrained to an `object` type. This would make it possible to detect conflicting constraints on *definition* of a generic type/routine (instead of during parameter type matching).
